arcgis-ios-sdk-samples/Maps/Map reference scale/MapReferenceScaleSettingsViewController.swift

```diff
@@ -0,0 +1,226 @@
+//
+// Copyright © 2019 Esri.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import UIKit
+import ArcGIS
+
+/// The delegate of a `MapReferenceScaleSettingsViewController`.
+protocol MapReferenceScaleSettingsViewControllerDelegate: AnyObject {
+    /// Tells the delegate that the user changed the map scale.
+    ///
+    /// - Parameter controller: The controller sending the message.
+    func mapReferenceScaleSettingsViewControllerDidChangeMapScale(_ controller: MapReferenceScaleSettingsViewController)
+    /// Tells the delegate that the user finished changing settings.
+    ///
+    /// - Parameter controller: The controller sending the message.
+    func mapReferenceScaleSettingsViewControllerDidFinish(_ controller: MapReferenceScaleSettingsViewController)
+}
+
+/// A view controller that provides an interface for adjusting the reference
+/// scale and scale of a map.
+class MapReferenceScaleSettingsViewController: UITableViewController {
+    /// The map whose settings should be adjusted.
+    var map: AGSMap!
+    /// The scale of the map. The default it `0`.
+    var mapScale = 0.0
+    /// The delegate of the view controller.
+    weak var delegate: MapReferenceScaleSettingsViewControllerDelegate?
+    
+    @IBOutlet weak var referenceScaleLabel: UILabel!
+    @IBOutlet weak var referenceScalePickerView: UIPickerView!
+    @IBOutlet weak var layersCell: UITableViewCell!
+    
+    @IBOutlet weak var mapScaleLabel: UILabel!
+    @IBOutlet weak var setMapScaleToReferenceScaleButton: UITableViewCell!
+    
+    /// The formatter used to generate strings from scale values.
+    private let scaleFormatter: NumberFormatter = {
+        let numberFormatter = NumberFormatter()
+        numberFormatter.numberStyle = .decimal
+        numberFormatter.maximumFractionDigits = 0
+        return numberFormatter
+    }()
+    /// The scale values displayed in the reference scale picker view.
+    private let possibleReferenceScales = [500_000, 250_000, 100_000, 50_000]
+    
+    /// Returns a string containing the formatted value of the provided scale.
+    ///
+    /// - Parameter scale: A scale value.
+    /// - Returns: A string.
+    func string(fromScale scale: Double) -> String {
+        return String(format: "1:%@", scaleFormatter.string(from: scale as NSNumber)!)
```

Comment:
can there ever be a case where `scaleFormatter.string` can return nil?  I know that sometimes a map `scale` will return `nan` if the map isn't loaded (or other circumstances), so would that cause problems with this?

I know in general we want to avoid force-unwrapping...

Reply:
> can there ever be a case where `scaleFormatter.string` can return nil?

Good question. I haven't found a case where it does:

```swift
numberFormatter.string(from: Double(0) as NSNumber) // "0"
numberFormatter.string(from: Double.nan as NSNumber) // "NaN"
numberFormatter.string(from: Double.infinity as NSNumber) // "+∞"
```

I suspect it returns an optional for the sake of Objective-C. I tested it and `-[NSNumberFormatter stringFromNumber:]` will return `nil` when passed `nil`. Obviously that isn't possible from Swift.